src/main/java/com/myFirstProject/myFirstProject/service/BasketServiceImpl.java

```diff
@@ -34,23 +38,46 @@
     @Autowired
     private ReqConverterService reqConverterService;
 
+    @Autowired
+    private PromoCodeRepository promoCodeRepository;
+
     @Override
     @Transactional
     public Long saveBasket(BasketReq basketReq) {
         logger.info("Get order to update {}", basketReq);
         sumValidation(basketReq);
+        PromoCode promoCode = getPromoCodeFromDB(basketReq);
         Basket order = reqConverterService.convert(basketReq);
+        order.setPromoCode(promoCode);
         LocalDateTime time = LocalDateTime.now();
         order.setExpiredTime(time.plusDays(1L));
-        Basket order1 = basketRepository.save(order);
-        Long id = order1.getId();
+        Basket savedOrder = basketRepository.save(order);
+        Long id = savedOrder.getId();
         logger.info("Basket saved with id {}", id);
-        accountService.payForArticles(basketReq.getSum(), basketReq.getUser().getId());
+        accountService.payForArticles(basketReq.getSum(), basketReq.getUser().getId(), promoCode);
+        promoCode.setValid(false);
         logger.info(String.format("%s was payed for order", basketReq.getSum()));
 
         return id;
     }
 
+    private PromoCode getPromoCodeFromDB(BasketReq basketReq) {
+        PromoCode promoCode = getPromoCode(basketReq);
+        if (promoCode == null) {
+            throw new PromoCodeNotFoundException(String.format("Promo code with id - %s not found", basketReq.getPromoCodeId()));
+        } else if (!promoCode.getValid()) {
+            throw new PromoCodeNotValidException(String.format("PromoCode with id - %s have been used", basketReq.getId()));
+        } else if (promoCode.getExpired().compareTo(LocalDateTime.now()) < 0) {
+            throw new PromoCodeNotValidException(String.format("PromoCode with id - %s expired", basketReq.getId()));
+        }
+        return promoCode;
+    }
+
+    private PromoCode getPromoCode(BasketReq basketReq) {
+        Optional<PromoCode> code = promoCodeRepository.findById(basketReq.getPromoCodeId());
+        return code.get();
```

Comment:
wrong style for Optional class:
 code.get(); - it provide NPE when PromoCode is absent in DB. This code can't work at all.

right solutions:
1) return option and then check, if exist
2) use getOrDefault or getOrThrow construction to avoid NPE